llvm/lib/SYCLLowerIR/SYCLDeviceRequirements.cpp

```diff
@@ -0,0 +1,136 @@
+//===----- SYCLDeviceRequirements.cpp - collect data for used aspects ----=-==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "llvm/SYCLLowerIR/SYCLDeviceRequirements.h"
+
+#include "llvm/ADT/SmallString.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/IR/Module.h"
+#include "llvm/SYCLLowerIR/ModuleSplitter.h"
+#include "llvm/Support/PropertySetIO.h"
+
+#include <set>
+#include <vector>
+
+using namespace llvm;
+
+static int64_t ExtractSignedIntegerFromMDNodeOperand(const MDNode *N,
+                                                     unsigned OpNo) {
+  Constant *C = cast<ConstantAsMetadata>(N->getOperand(OpNo).get())->getValue();
+  return C->getUniqueInteger().getSExtValue();
+}
+static uint64_t ExtractUnsignedIntegerFromMDNodeOperand(const MDNode *N,
+                                                        unsigned OpNo) {
+  Constant *C = cast<ConstantAsMetadata>(N->getOperand(OpNo).get())->getValue();
+  return C->getUniqueInteger().getZExtValue();
+}
+static llvm::StringRef ExtractStringFromMDNodeOperand(const MDNode *N,
+                                                      unsigned OpNo) {
+  MDString *S = cast<llvm::MDString>(N->getOperand(OpNo).get());
+  return S->getString();
+}
+
+SYCLDeviceRequirements
+llvm::getSYCLDeviceRequirements(const module_split::ModuleDesc &MD) {
+  SYCLDeviceRequirements Reqs;
+  // Process all functions in the module
+  for (const Function &F : MD.getModule()) {
+    if (auto *MDN = F.getMetadata("sycl_used_aspects")) {
+      for (size_t I = 0, E = MDN->getNumOperands(); I < E; ++I) {
+        auto Val = ExtractSignedIntegerFromMDNodeOperand(MDN, I);
+        // Don't put internal aspects (with negative integer value) into the
+        // requirements, they are used only for device image splitting.
+        if (Val >= 0)
+          Reqs.Aspects.insert(Val);
+      }
+    }
+
+    if (auto *MDN = F.getMetadata("sycl_fixed_targets")) {
+      for (size_t I = 0, E = MDN->getNumOperands(); I < E; ++I) {
+        auto Val = ExtractUnsignedIntegerFromMDNodeOperand(MDN, I);
+        Reqs.FixedTarget.insert(Val);
+      }
+    }
+
+    if (auto *MDN = F.getMetadata("reqd_work_group_size")) {
+      llvm::SmallVector<uint64_t, 3> NewReqdWorkGroupSize;
+      for (size_t I = 0, E = MDN->getNumOperands(); I < E; ++I)
+        NewReqdWorkGroupSize.push_back(
+            ExtractUnsignedIntegerFromMDNodeOperand(MDN, I));
+      if (!Reqs.ReqdWorkGroupSize.has_value())
+        Reqs.ReqdWorkGroupSize = NewReqdWorkGroupSize;
```

Comment:
I think we need `else` branch here to validate consistency of multiple "reqd_work_group_size" requirements.

Reply:
If we do this, then this test fails:

https://github.com/intel/llvm/blob/93ce40a7e0bfad3630eaeb5ecffd3f197b19055c/llvm/test/tools/sycl-post-link/device-code-split/per-reqd-wg-size-split-3.ll#L13-L17

Since splitting was disabled, there will be multiple `reqd_work_group_size` attributes in a module, and then it'll run into an assert if we placed on there. I think the surrounding code that calls `getSYCLDeviceRequirements` would need to be updated to not compute device requirements when splitting is disabled if we were to add an assert here.